src/vf2_layout.rs

```diff
@@ -0,0 +1,121 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use indexmap::IndexMap;
+
+use numpy::{PyReadonlyArray1, PyReadonlyArray2};
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use rayon::prelude::*;
+
+use crate::nlayout::NLayout;
+
+const PARALLEL_THRESHOLD: usize = 50;
+
+/// Score a given circuit with a layout applied
+#[pyfunction]
+#[pyo3(
+    text_signature = "(bit_list, edge_list, error_matrix, layout, strict_direction, run_in_parallel, /)"
+)]
+pub fn score_layout(
+    bit_list: PyReadonlyArray1<i32>,
+    edge_list: IndexMap<[usize; 2], i32>,
+    error_matrix: PyReadonlyArray2<f64>,
+    layout: &NLayout,
+    strict_direction: bool,
+    run_in_parallel: bool,
+) -> PyResult<f64> {
+    let bit_counts = bit_list.as_slice()?;
+    let err_mat = error_matrix.as_array();
+    let mut fidelity = if edge_list.len() < PARALLEL_THRESHOLD || !run_in_parallel {
+        edge_list
+            .iter()
+            .filter_map(|(index_arr, gate_count)| {
+                let mut error = err_mat[[
+                    layout.logic_to_phys[index_arr[0]],
+                    layout.logic_to_phys[index_arr[1]],
+                ]];
+                if !strict_direction && error.is_nan() {
+                    error = err_mat[[
+                        layout.logic_to_phys[index_arr[1]],
+                        layout.logic_to_phys[index_arr[0]],
+                    ]];
+                }
+                if !error.is_nan() {
+                    Some((1. - error).powi(*gate_count))
+                } else {
+                    None
+                }
+            })
+            .reduce(|a, b| a * b)
+            .unwrap_or(1.)
```

Comment:
Oh, yeah there is I forgot about that, that'll simplify this iterator chain.  (internally it looks like it uses `fold()` instead of `reduce()`) It should work with the parallel path too.

Reply:
Done in https://github.com/Qiskit/qiskit-terra/pull/9026/commits/295a8294f31ecdbe7946a19c695cc3991cd934c2